### PR TITLE
Adding support for Acknowledge, Beacons, Data and MAC commands

### PR DESCRIPTION
**Pull request by blueluna**
This PR adds support for encoding and decoding Beacons and MAC commands. This is my first time working with 802.15.4 so I might not understand all the concepts. But here goes.

There probably are opportunities for better naming in some part of the code (for example FrameContent?).

Testing done.
I have decoded `Acknowledge`, `Beacon`, `Command::AssociationRequest`, `Command::AssociationResponse`, `Command::BeaconRequest`, `Command::DataRequest`, `Data` from a coordinator or other devices in a PAN. I have encoded `Command::AssociationRequest`, `Command::BeaconRequest`, `Command::DataRequest` which has been understood (I believe) by a coordinator.

### CODE REVIEW

File: src/mac/beacon.rs [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+impl Beacon {
+    /// Decode beacon frame from byte  buffer
+    ///
+    /// # Returns
+    ///
+    /// Returns [`Beacon`] and the number of bytes used are returned
+    ///
+    /// # Errors
+    ///
+    /// This function returns an error, if the bytes either don't are enough or
```
Comment (hannobraun) [inline src/mac/beacon.rs]:
```suggestion
    /// This function returns an error, if there aren't enough bytes or
```

File: src/mac/beacon.rs [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+    /// # Panics
+    ///
+    /// Panics if the buffer is not long enough to hold the frame.
+    ///
+    pub fn encode(&self, buf: &mut [u8]) -> usize {
+        let size = self.superframe_spec.encode(buf);
+        let mut offset = size;
+        let size = self.guaranteed_time_slot_info.encode(&mut buf[offset..]);
+        offset += size;
+        let _ = self.pending_address.encode(&mut buf[offset..]);
```
Comment (hannobraun) [inline src/mac/beacon.rs]: Can you please explain why the second size is ignored, and the first size added twice? Is this intentional?

File: src/mac/command.rs [hunk 3]
```diff
@@ -0,0 +1,817 @@
...
+extended_enum!(
+    /// Association Status
+    AssociationStatus, u8,
+    /// Successful
+    Successful => 0x00,
+    /// Network (PAN) at capacity
+    NetworkAtCapacity => 0x01,
+    /// Access to PAN denied
+    AccessDenied => 0x02,
+    /// ???
```
Comment (hannobraun) [inline src/mac/command.rs]: This comment is not very informative :-)

File: src/mac/command.rs [hunk 4]
```diff
@@ -0,0 +1,817 @@
...
+impl CoordinatorRealignmentData {
+    /// Decode coordinator re-alignment data from byte buffer
+    ///
+    /// # Returns
+    ///
+    /// Returns [`CoordinatorRealignmentData`] and the number of bytes used are returned
+    ///
+    /// # Errors
+    ///
+    /// This function returns an error, if the bytes either don't are enough or
```
Comment (hannobraun) [inline src/mac/command.rs]:
```suggestion
    /// This function returns an error, if there aren't enough bytes or
```

File: src/mac/mod.rs [hunk 5]
```diff
@@ -0,0 +1,21 @@
+//! Partial implementation of the IEEE 802.15.4 MAC layer
+
+mod frame;
```
Comment (hannobraun) [inline src/mac/mod.rs]:
```suggestion
pub mod frame;
```
Not necessary, but doesn't hurt and it consistent with the other modules.

File: src/mac/mod.rs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+pub mod command;
+pub mod beacon;
+
+pub use frame::{
```
Comment (hannobraun) [inline src/mac/mod.rs:7]: Since this is a breaking change anyway, we might consider not re-exporting all of these (or any of these). Not sure.

**Review by hannobraun — COMMENTED** [review]
Thank you @blueluna! I don't have the standard, and I don't have any hardware running other implementations of the protocol, so my ability to review and test this is limited. Still, I've looked over the code and left some comments.

Let me know what you think about the comments. Once they have been resolved, I'm happy to merge.

File: src/mac/beacon.rs (hunk 2 above)
Comment (blueluna) [reply src/mac/beacon.rs]: No, this seems like an error. The size from encoding pending address should probably be used,
```suggestion
        let size = self.pending_address.encode(&mut buf[offset..]);
```
I will investigate.

File: src/mac/command.rs (hunk 3 above)
Comment (blueluna) [reply src/mac/command.rs]: No, I will come up with something better.

File: src/mac/mod.rs (hunk 6 above)
Comment (blueluna) [reply src/mac/mod.rs:7]: I do not have a strong opinion. I might prefer it this way, since I think these enums and structs will be used the most.

Comment (hannobraun) [reply src/mac/mod.rs:7]: I don't feel strongly either. Let's leave it as is then, for now.

**Comment by hannobraun** [comment]
Okay, this looks good to merge. Thank you again, @blueluna!

**Pull request merged**